daemon/daemon.go

```diff
@@ -1040,23 +1044,56 @@ func (daemon *Daemon) TagImage(newTag reference.Named, imageName string, force b
 	return daemon.tagStore.Add(newTag, imageID, force)
 }
 
+func writeDistributionProgress(cancelFunc func(), outStream io.Writer, progressChan <-chan xfer.Progress) {
+	sf := streamformatter.NewJSONStreamFormatter()
+	for prog := range progressChan {
+		if prog.Message != "" {
+			formatted := sf.FormatStatus(prog.ID, prog.Message)
+			if _, err := outStream.Write(formatted); err != nil {
+				cancelFunc()
+			}
+		} else {
+			jsonProgress := jsonmessage.JSONProgress{Current: prog.Current, Total: prog.Total}
+			formatted := sf.FormatProgress(prog.ID, prog.Action, &jsonProgress)
+			if _, err := outStream.Write(formatted); err != nil {
+				cancelFunc()
+			}
+		}
+	}
+}
+
 // PullImage initiates a pull operation. image is the repository name to pull, and
 // tag may be either empty, or indicate a specific tag to pull.
 func (daemon *Daemon) PullImage(ref reference.Named, metaHeaders map[string][]string, authConfig *cliconfig.AuthConfig, outStream io.Writer) error {
+	// Include a buffer so that slow client connections don't affect
+	// transfer performance.
+	progressChan := make(chan xfer.Progress, 100)
+
+	writesDone := make(chan struct{})
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	go func() {
+		writeDistributionProgress(cancelFunc, outStream, progressChan)
+		close(writesDone)
+	}()
+
 	imagePullConfig := &distribution.ImagePullConfig{
 		MetaHeaders:     metaHeaders,
 		AuthConfig:      authConfig,
-		OutStream:       outStream,
+		ProgressChan:    progressChan,
 		RegistryService: daemon.RegistryService,
 		EventsService:   daemon.EventsService,
 		MetadataStore:   daemon.distributionMetadataStore,
-		LayerStore:      daemon.layerStore,
 		ImageStore:      daemon.imageStore,
 		TagStore:        daemon.tagStore,
-		Pool:            daemon.distributionPool,
+		DownloadManager: daemon.downloadManager,
 	}
 
-	return distribution.Pull(ref, imagePullConfig)
+	err := distribution.Pull(ctx, ref, imagePullConfig)
+	close(progressChan)
```

Comment:
The trick here is to make sure the listen side never attempts to close the channel. Patterns like this work fine:

```
select {
case <-closed:
    return ErrClosed
default:
    close(closed)
}
```

The key is ensuring ownership on the channel. Since this code owns the channel and delegates sending and receiving, I don't think we are running afoul of the best practice. We would be in the close was done in `writeDistributionProgress`. We only need to be sure that the channel can only ever be closed once and once it is closed, no sends can occur.

While sending on a closed channel panics, I don't this applies when using a select statement, since the operation cannot proceed. (I may be wrong here).